demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+// Copyright (C) 2019 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "cpu_monitor.h"
+#ifdef WIN32
+#include <tchar.h>
+
+namespace {
+std::size_t getNCores() {
+    SYSTEM_INFO sysinfo;
+    GetSystemInfo(&sysinfo);
+    return sysinfo.dwNumberOfProcessors;
+}
+}
+
+CpuMonitor::CpuMonitor() : nCores{getNCores()}, lastEnabled{false}, historyEnabled{false}, samplesNumber{0} {};
+
+void CpuMonitor::openQuery() {
+    PDH_STATUS status;
+
+    status = PdhOpenQuery(NULL, NULL, &query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhOpenQuery() failed");
+    }
+
+    for (std::size_t i = 0; i < nCores; ++i)
+    {
+        TCHAR szCounterName[MAX_PATH];
+        _stprintf_s(szCounterName, sizeof(szCounterName), TEXT("\\Processor(%u)\\%% Processor Time"), i);
+
+        PDH_HCOUNTER counter;
+
+        status = PdhAddCounter(query, szCounterName, 0, &counter);
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhAddCounter() failed");
+        }
+
+        status = PdhSetCounterScaleFactor(counter, -2); // scale counter to [0, 1]
+        if (ERROR_SUCCESS != status)
+        {
+            throw std::runtime_error("PdhSetCounterScaleFactor() failed");
```

Comment:
`system_error` would be a better fit here, as it would allow you to include the error code in the exception data.

Reply:
updated